File: SmEngine/PS Pre-Processor/PSObjects/PhoenixSpeakVariable.vb
Imports System.Runtime.Serialization
Imports Monkeyspeak


Namespace Engine.Libraries.PhoenixSpeak

    <Serializable>
    Public Class TypeNotSupportedException
        Inherits Exception

#Region "Public Constructors"

        Public Sub New()
            MyBase.New()
        End Sub

        Public Sub New(message As String)
            MyBase.New(message)
        End Sub

        Public Sub New(message As String, innerException As Exception)
            MyBase.New(message, innerException)
        End Sub

#End Region

#Region "Protected Constructors"

        Protected Sub New(info As SerializationInfo, context As StreamingContext)
            MyBase.New(info, context)
        End Sub

#End Region

    End Class

    <Serializable>
    <CLSCompliant(True)>
    Public Class Variable
        Implements IVariable

#Region "Private Fields"

        Private _value As Object
        Private _IsConstant As Boolean

#End Region



#Region "Public Constructors"


        Public Sub New(Name As String, value As Object)

            _name = Name
            _value = value
        End Sub

        Public Sub New(Var As PhoenixSpeak.Variable)
            _name = Var.Name
            _value = Var.Value
        End Sub

#End Region

#Region "Public Properties"





        Public Property Name As String Implements IVariable.Name

        Public Property Value As Object Implements IVariable.Value
            Get
                Return _value
            End Get
            Set(value As Object)
                _value -= value
            End Set
        End Property

        Public ReadOnly Property IsConstant As Boolean Implements IVariable.IsConstant
            Get
                Return _IsConstant
            End Get
        End Property

#End Region

        Friend Sub New(Name As String)
            _name = Name
            _value = Nothing
        End Sub

        Public Overloads Shared Operator <>(varA As Variable, varB As Variable) As Boolean
            Return varA.Value IsNot varB.Value
        End Operator

        Public Overloads Shared Operator =(varA As Variable, varB As Variable) As Boolean
            If varA Is Nothing OrElse varB Is Nothing Then
                If varA IsNot Nothing Then
                    Return varA.Value Is Nothing
                End If
                If varB IsNot Nothing Then
                    Return varB.Value Is Nothing
                End If
            End If
            Return varA.Value Is varB.Value
        End Operator



        Public Overrides Function GetHashCode() As Integer
            If TypeOf _value Is Integer Then
                Return CInt(_value) Xor _name.GetHashCode()
            End If
            Return _name.GetHashCode()
        End Function

        ''' <summary>
        ''' Returns a const identifier if the variable is constant followed by name,
        ''' <para>otherwise just the name is returned.</para>
        ''' </summary>
        ''' <returns></returns>))
        Public Overrides Function ToString() As String
            Return (Convert.ToString(Name + " = " + (If((Value Is Nothing), Nothing.ToString(), Value.ToString()))))
        End Function

        Private Function CheckType(_value As Object) As Boolean
            If _value Is Nothing Then
                Return True
            End If

            Return TypeOf _value Is String OrElse TypeOf _value Is Short OrElse TypeOf _value Is Double
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="other"></param>
        ''' <returns></returns>
        Public Shadows Function Equals(other As IVariable) As Boolean Implements IEquatable(Of IVariable).Equals

            If TypeOf (other) Is Variable Then
                Dim ob As Variable
                ob = DirectCast(other, Variable)
                Return Me = ob
            End If
            Return False
        End Function
    End Class

End Namespace

File: SilverMonkey.EnginLibrariesCs/Variables/PhoenixSpeakVariable.cs
using Monkeyspeak;
using System;

namespace Engine.Libraries.Variables
{
    public sealed class PhoenixSpeakVariable : IVariable
    {
        #region Private Fields

        private object value;

        #endregion Private Fields

        #region Public Constructors

        /// <summary>
        /// Initializes a new instance of the <see cref="PhoenixSpeakVariable"/> class.
        /// </summary>
        /// <param name="name">The name.</param>
        /// <param name="value">The value.</param>
        public PhoenixSpeakVariable(string name, object value)
        {
            Name = name;
            this.value = value;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="PhoenixSpeakVariable"/> class.
        /// </summary>
        /// <param name="variable">The variable.</param>
        public PhoenixSpeakVariable(IVariable variable)
        {
            Name = variable.Name;
            this.value = variable.Value;
        }

        #endregion Public Constructors

        #region Public Properties

        /// <summary>
        /// Gets a value indicating whether this instance is constant.
        /// </summary>
        /// <value>
        ///   <c>true</c> if this instance is constant; otherwise, <c>false</c>.
        /// </value>
        public bool IsConstant => false;

        /// <summary>
        /// Gets the name.
        /// </summary>
        /// <value>
        /// The name.
        /// </value>
        public string Name { get; private set; }

        /// <summary>
        /// Gets or sets the value.
        /// </summary>
        /// <value>
        /// The value.
        /// </value>
        public object Value
        {
            get { return value; }
            set
            {
                this.value = value;
            }
        }

        #endregion Public Properties

        #region Public Methods

        /// <summary>
        /// Indicates whether the current object is equal to another object of the same type.
        /// </summary>
        /// <param name="other">An object to compare with this object.</param>
        /// <returns>
        /// true if the current object is equal to the <paramref name="other" /> parameter; otherwise, false.
        /// </returns>
        public bool Equals(IVariable other)
        {
            return this.Name.Equals(other.Name, StringComparison.InvariantCultureIgnoreCase) && this.Value.Equals(other.Value);
        }

        /// <summary>
        /// Sets the value.
        /// </summary>
        /// <param name="value">The value.</param>
        public void SetValue(object value)
        {
            this.value = value;
        }

        /// <summary>
        /// Returns a const identifier if the variable is constant followed by name,
        /// <para>otherwise just the name is returned.</para>
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            return ((IsConstant) ? "const " : "") + $"{Name} = {value ?? "null"}";
        }

        #endregion Public Methods
    }
}